### ISSUE TITLE
S2583 - false positive

### ISSUE THREAD

**Issue by JohnyWS** [SonarSource/sonar-dotnet#2861]
### Description

We've encountered a false positive when running SQ on PR, that we cannot reproduce with SonarLint. See code snippet for details.

As far as I understand it, it should be this code that errors in parsing it, but could not immediately locate the error though: https://github.com/SonarSource/sonar-dotnet/blob/master/sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConditionEvaluatesToConstant.cs

### Repro steps

The following is the snippet that fails, on the last line where `!writeDeny` is marked as always true. I've simplified the code a little bit, but hopefully only removing guards that weren't necessary to reproduce it.
```
var info = new DirectoryInfo(Environment.CurrentDirectory);

var accessRules = info.GetAccessControl()?.GetAccessRules(true, true, typeof(System.Security.Principal.SecurityIdentifier));

bool writeAllow = false, writeDeny = false;
foreach (FileSystemAccessRule rule in accessRules)
{
  if ((FileSystemRights.Write & rule.FileSystemRights) != FileSystemRights.Write) continue;

  if (rule.AccessControlType == AccessControlType.Allow) writeAllow = true;
  else if (rule.AccessControlType == AccessControlType.Deny) writeDeny = true;
}

writeAllow && !writeDeny ? DoSomething() : DoSomethingElse();
```

### Expected behavior

All good

### Actual behavior

`!writeDeny` was marked as in violation of S2583

### Known workarounds

None

### Related information

* SonarC# Version: 7.9.1.27448
* Visual Studio Version: Unrelated, was in PR
* If running through the Scanner for MSBuild, its version: dotnet.exe 2.131.0

**Comment by JohnyWS** [comment]
Any further info needed for you to give this a look? :)

**Comment by andrei-epure-sonarsource** [comment]
Hi @JohnyWS, thanks for reaching out. There is no SonarCSharp version 7.9.1.27448, you probably refer to the SonarQube version (the SonarCSharp version released a year ago is  [7.9.1.7622](https://github.com/SonarSource/sonar-dotnet/releases/tag/7.9.1.7622)).

In SonarCSharp 7.16 we fixed a lot of FPs (this is why in SonarLint you cannot repro, probably the SonarLint is up to date), so I recommend to update inside the SQ Marketplace the version of the SonarCSharp plugin to the latest ([8.1.0.13383](https://github.com/SonarSource/sonar-dotnet/releases/tag/8.1.0.13383)) and let us know if the FPs are still there.

**Comment by JohnyWS** [comment]
Finally got around to updating it (resulted in other update issues, but got it fixed now) - so yes, it was old plugins we were using, and we did not have rulessets set up to inherit correctly, so that improved our setup tremendously. Thanks! :)
